README.md
# Adonis Translatable

AdonisJS Lucid internationalization models.

## Installation

**npm**

```
npm install @emilioforrer/adonis-translatable
```

**yarn**

```
yarn add @emilioforrer/adonis-translatable
```


Add the provider to your start/app.js file:

```js
const providers = [
   // ...
   "@emilioforrer/providers/TranslatableProvider",
   // ...
 ];
```

## How it works

Basically this addon creates a new relationship called `translations` to main model and delegates it's translatable attributes to the translation model.

So in order to work, you need to create a new model translations for your main model, add a trait and configure it.

## How to use

Let's say we have a model `App/Models/Role` and want to translate the name and description attributes, in order to do that follow the next steps.

1. Let's create the translation table for the model `App/Models/Role`.

```js
schema.createTable('roles_translations', (table) => {
  table.increments()
  table.timestamps()
  /*
   * Translation fields
   */
  table.string('name').nullable()
  table.text('description').nullable()
  table.integer('role_id').unsigned()
  table.string('language_code')
})
```
2. Now let's create the translation model `App/Models/RoleTranslation`

```js
const Model = app.use('Model')
class RoleTranslation extends Model {
  static get table () {
    return 'roles_translations'
  }
}
```

3. Next, we need to add a `trait` to the main model (`App/Models/Role`) and configure the options

```js
const Model = app.use('Model')
class Role extends Model {
  static boot () {
    super.boot()
    this.addTrait('Translatable', {
      className: 'App/Models/RoleTranslation',
      attributes: ['name', 'description']
    })
  }
}
```

## Translatable trait options

* **className:** The name of the translation model. Defaults to: `''`.
* **primaryKey:** This is the primary key that will be used to create the `translation` relationship. Defaults to: `Model.primaryKey`.
* **foreignKey:** This is the foreign key that will be used to create the `translation` relationship. Defaults to: `${singlularized_table_name}_id`.
* **attributes:** The list of attributes that we wish to be delegated to the translation model.

## Development

Great! If you are planning to contribute to the framework, make sure to adhere to following conventions, since a consistent code-base is always joy to work with.

Run the following command to see list of available npm scripts.

**npm**

```bash
npm run
```

**yarn**

```bash
yarn run
```

### Tests & Linting

1. Lint your code using standardJs. Run `yarn run lint` command to check if there are any linting errors.
2. Make sure you write tests for all the changes/bug fixes.
3. Run `yarn run test` to lint and run your tests.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.


## [1.0.1] - 2018-11-16

### Fixed
- Issue #1: Translation attribute cannot bind to main model when call .toJSON()
 

## [1.0.0] - 2018-05-06

### Added
- Everything, initial release

test/setup.js
'use strict'

/**
 * adonis-persona
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/

process.env.SILENT_ENV = true

const path = require('path')
const { registrar, ioc } = require('@adonisjs/fold')
const { setupResolver, Helpers } = require('@adonisjs/sink')

module.exports = {
  wire: async function () {
    setupResolver()
    ioc.bind('Adonis/Src/Helpers', () => new Helpers(path.join(__dirname, '..', 'app')))

    await registrar.providers([
      '@adonisjs/framework/providers/AppProvider',
      '@adonisjs/lucid/providers/LucidProvider',
      '@adonisjs/validator/providers/ValidatorProvider',
      '@adonisjs/antl/providers/AntlProvider',
      path.join(__dirname, '../providers/TranslatableProvider')
    ]).registerAndBoot()

    ioc.singleton('App/Models/Role', (app) => {
      const Model = app.use('Model')
      class Role extends Model {
        static boot () {
          super.boot()
          this.addTrait('Translatable', {
            className: 'App/Models/RoleTranslation',
            attributes: ['name', 'description']
          })
        }
      }
      Role._bootIfNotBooted()
      return Role
    })

    ioc.singleton('App/Models/RoleTranslation', (app) => {
      const Model = app.use('Model')
      class RoleTranslation extends Model {
        static get table () {
          return 'roles_translations'
        }
      }
      RoleTranslation._bootIfNotBooted()
      return RoleTranslation
    })
  },

  async migrateUp () {
    await use('Database').schema.createTable('languages', (table) => {
      table.string('name')
      table.string('code').primary()
      table.timestamps()
    })

    await use('Database').schema.createTable('roles', (table) => {
      table.increments()
      table.string('slug').notNullable().unique()
      table.timestamps()
    })

    await use('Database').schema.createTable('roles_translations', (table) => {
      table.increments()
      table.timestamps()

      table.string('name').nullable()
      table.text('description').nullable()
      /*
       * Translation fields
       */
      table.integer('role_id').unsigned()
      table.string('language_code')
      // table.foreign(['role_id', 'language_code'])
      //   .onDelete('CASCADE')
      //   .onUpdate('RESTRICT')
      //   .references(['roles.id', 'languages.code'])
    })
  },

  async migrateDown () {
    await use('Database').schema.dropTableIfExists('languages')
    await use('Database').schema.dropTableIfExists('roles')
    await use('Database').schema.dropTableIfExists('roles_translations')
  }
}


test/antldb.spec.js
'use strict'

const test = require('japa')
const setup = require('./setup')
const moment = require('moment')

function getRole () {
  return use('App/Models/Role')
}

test.group('Antldb', (group) => {
  group.before(async () => {
    await setup.wire()
    await setup.migrateDown()
    await setup.migrateUp()
  })

  group.beforeEach(async () => {
    await use('Database').beginGlobalTransaction()
  })

  group.afterEach(async () => {
    // await use('Database').rollbackGlobalTransaction()
  })

  group.after(async () => {
    use('Database').close()
  })

  test('test role trait', async (assert) => {
    let role = await getRole().create({
      slug: 'admin',
      created_at: moment().format('YYYY-MM-DD HH:mm:ss'),
      updated_at: moment().format('YYYY-MM-DD HH:mm:ss')
    })
    role.name = 'Administrator'
    assert.deepEqual(role.name, 'Administrator')
    await role.save()
    let translation = await role.translation().fetch()
    assert.deepEqual(translation.name, 'Administrator')
  })
})


test/translatable.spec.js
'use strict'

const test = require('japa')
const setup = require('./setup')
const moment = require('moment')

function getRole () {
  return use('App/Models/Role')
}

test.group('Antldb', (group) => {
  group.before(async () => {
    await setup.wire()
    await setup.migrateDown()
    await setup.migrateUp()
  })

  group.beforeEach(async () => {
    await use('Database').beginGlobalTransaction()
  })

  group.afterEach(async () => {
    // await use('Database').rollbackGlobalTransaction()
  })

  group.after(async () => {
    use('Database').close()
  })

  test('test role trait', async (assert) => {
    let role = await getRole().create({
      slug: 'admin',
      created_at: moment().format('YYYY-MM-DD HH:mm:ss'),
      updated_at: moment().format('YYYY-MM-DD HH:mm:ss')
    })
    role.name = 'Administrator'
    assert.deepEqual(role.name, 'Administrator')
    await role.save()
    let translation = role.getRelated('translation')
    assert.deepEqual(translation.name, 'Administrator')

    role = await getRole().query().first()
    assert.deepEqual(role.name, 'Administrator')

    const objectRole = role.toObject()

    assert.deepEqual(objectRole.name, 'Administrator')

    const jsonRole = role.toObject()

    assert.deepEqual(jsonRole.name, 'Administrator')

    let roles = await getRole().all()
    for (let record of roles.rows) {
      assert.deepEqual(record.name, 'Administrator')
    }
  })
})


providers/TranslatableProvider.js
'use strict'

/**
 * @Author: Emilio Forrer <emilioforrer>
 * @Date:   2018-05-02T22:10:11-06:00
 * @Email:  dev.emilio.forrer@gmail.com
 * @Filename: TranslatableProvider.js
 * @Last modified by:   emilioforrer
 * @Last modified time: 2018-05-06T11:49:42-06:00
 * @License: For the full license information, please view the LICENSE
 * @Copyright: (c) Emilio Forrer
 */

const { ServiceProvider } = use('@adonisjs/fold') // require.main.require('@adonisjs/fold') as for https://forum.adonisjs.com/t/v-4-create-serviceproiver-outside-app/128/19

class TranslatableProvider extends ServiceProvider {
  _registerMiddlewares () {

  }

  _registerTraits () {
    this.app.bind('App/Models/Traits/Translatable', (app) => {
      const Translatable = require('../src/Models/Traits/TranslatableTrait')
      return new Translatable()
    })
  }

  register () {
    this._registerMiddlewares()
    this._registerTraits()
  }

  boot () {
  }
}

module.exports = TranslatableProvider


src/Models/Traits/TranslatableTrait.js
/**
 * @Author: Emilio Forrer <emilioforrer>
 * @Date:   2018-05-06T11:45:54-06:00
 * @Email:  dev.emilio.forrer@gmail.com
 * @Filename: TranslatableTrait.js
 * @Last modified by:   emilioforrer
 * @Last modified time: 2018-05-06T11:49:04-06:00
 * @License: For the full license information, please view the LICENSE
 * @Copyright: (c) Emilio Forrer
 */

const _ = use('lodash')
const Antl = use('Antl')
const pluralize = use('pluralize')

class TranslatableTrait {
  async setLocalize (modelInstance, translatableAttributes) {
    for (let translatableAttribute of translatableAttributes) {
      modelInstance.$localizedAttributes[translatableAttribute] = modelInstance[translatableAttribute]
    }
  }

  async localize (modelInstance, translatableAttributes) {
    /* istanbul ignore else */
    if (modelInstance.$persisted) {
      let translation = modelInstance.getRelated('translation')
      /* istanbul ignore else */
      if (!translation) {
        await modelInstance.load('translation')
        translation = modelInstance.getRelated('translation')
      }
      /* istanbul ignore else */
      if (translation) {
        for (let translatableAttribute of translatableAttributes) {
          modelInstance[translatableAttribute] = translation[translatableAttribute]
        }
      }
    }
  }

  async register (Model, customOptions) {
    const defaultOptions = {
      className: '',
      primaryKey: Model.primaryKey,
      foreignKey: `${_.snakeCase(pluralize.singular(Model.table))}_id`,
      attributes: []
    }
    const options = _.extend({}, defaultOptions, customOptions)
    const that = this
    /* istanbul ignore else */
    if (_.isArray(options.attributes)) {
      let _instantiate = Model.prototype._instantiate

      // Model.translatableOptions = options.translatableAttributes

      const _toObject = Model.prototype.toObject

      Model.prototype.toObject = function () {
        let json = _toObject.call(this)
        for (let translatableAttribute of options.attributes) {
          json[translatableAttribute] = this[translatableAttribute]
        }
        return json
      }

      Model.prototype.translation = function () {
        return this.hasOne(
          options.className,
          options.primaryKey,
          options.foreignKey
        ).where('language_code', Antl._locale)
      }

      Model.prototype._instantiate = async function () {
        _instantiate.call(this)
        this.__setters__.push('$localizedAttributes')
        this.$localizedAttributes = {}
        for (let translatableAttribute of options.attributes) {
          this.__setters__.push(translatableAttribute)
        }
        // await that.localize(this, options.attributes)
        await that.setLocalize(this, options.attributes)
      }

      Model.addHook('afterFind', async (modelInstance) => {
        await that.localize(modelInstance, options.attributes)
      })

      Model.addHook('afterFetch', async (instances) => {
        for (const instance of instances) {
          await that.localize(instance, options.attributes)
        }
      })

      Model.addHook('beforeSave', async (instance) => {

      })

      Model.addHook('afterSave', async (instance) => {
        let translation = instance.getRelated('translation')
        if (!translation) {
          await instance.load('translation')
          translation = instance.getRelated('translation')
        }
        if (!translation) {
          let ClassName = use(options.className)
          translation = new ClassName()
          translation.language_code = Antl._locale
        }
        for (let translatableAttribute of options.attributes) {
          translation[translatableAttribute] = instance[translatableAttribute]
        }
        await instance.translation().save(translation)
        instance.$relations['translation'] = translation
      })
    } else {
      throw new Error('option attributes must return an array')
    }
  }
}

module.exports = TranslatableTrait
